fluid system overhaul

The4codeblocks opened the pull request:
autocrafter fluid support
creative fluid source
trashcan acting as fluid void
flow sensor acting as a pressure gauge (shows fluid type too)
multiple fluid types (no mixing, but there is displacement)
| has custom fluid registration
| updating existing pump and drains to support multiple fluids
| per fluid "finite" config

**(important detail: flow logic default set to `"pressure"` instead of `"classic"`)**

The4codeblocks committed:
trashcan voids water

The4codeblocks committed:
trashcan updates pipes

The4codeblocks committed:
trash can needs pipe logic now

The4codeblocks committed:
offsets

The4codeblocks committed:
pressure default

The4codeblocks committed:
creative water source

The4codeblocks committed:
fluid crafts

The4codeblocks committed:
autocrafter with fluids

The4codeblocks committed:
Add files via upload

The4codeblocks committed:
fluid crafts without core

The4codeblocks committed:
update crafts

The4codeblocks committed:
fix offsets

The4codeblocks committed:
flow sensor info

The4codeblocks committed:
don't global

The4codeblocks committed:
stuff

The4codeblocks committed:
add lava and defs

The4codeblocks committed:
Update autocrafter.lua

The4codeblocks committed:
remove unnecessary registration

The4codeblocks committed:
Add files via upload

The4codeblocks committed:
recipes

The4codeblocks committed:
Update crafts.lua

The4codeblocks committed:
enforcing load order

BuckarooBanzay commented on default_settings.lua:
```diff
@@ -48,7 +48,7 @@ pipeworks.toggles = {}
 -- or for rendering the pipes purely decorative.
 ]]
-pipeworks.toggles.pipe_mode = "classic"
+pipeworks.toggles.pipe_mode = "pressure"
```
how does that affect existing worlds?

The4codeblocks replied:
a lot of the things I'm doing *are* going to break things, but the old way things work is likely simple enough to auto-reinstate the original function

BuckarooBanzay commented:
Current luacheck warnings: (i removed the ones about missing `core` global)

```
Checking autocrafter.lua                          12 warnings

    autocrafter.lua:5:1: inconsistent indentation (SPACE followed by TAB)
    autocrafter.lua:17:1: line contains only whitespace
    autocrafter.lua:24:1: line contains only whitespace
    autocrafter.lua:181:1: line contains only whitespace
    autocrafter.lua:194:1: line contains only whitespace

 Checking common.lua                               1 warning

    common.lua:223:9: unused variable inv_size

Checking trashcan.lua                             2 warnings

    trashcan.lua:48:2: value assigned to field after_place_node is overwritten on line 53 before use
    trashcan.lua:49:2: value assigned to field after_dig_node is overwritten on line 56 before use
```

sounds good imo, didn't test that yet though

The4codeblocks committed:
merge on_place and on_dig (contains leak from wip)

The4codeblocks committed:
fix luacheck again

The4codeblocks committed:
remove space-only lines

The4codeblocks committed:
Update .luacheckrc

The4codeblocks committed:
generalize flag

The4codeblocks committed:
Update init.lua

The4codeblocks committed:
actual multifluid support

The4codeblocks committed:
casewise finity

The4codeblocks committed:
try dehardcoding possibly useless code

The4codeblocks committed:
sensor show type

The4codeblocks committed:
semantic changes (water → fluid)

The4codeblocks committed:
support for existing stuff for multitype

The4codeblocks committed:
a WHOLE lot

The4codeblocks committed:
fix, bad copy

The4codeblocks committed:
fix allowing fluid outs to change stored type for some reason

The4codeblocks committed:
last commit summary was a bit vague

The4codeblocks committed:
Update pressure_logic.txt

The4codeblocks committed:
did that not fix? hmm

The4codeblocks commented:
feature set in desc updated

The4codeblocks committed:
oops, forgot to remove that logger

The4codeblocks committed:
oops, forgot to remove that logger

The4codeblocks committed:
multifluid source

The4codeblocks committed:
check for definitions 1st

The4codeblocks committed:
shut luacheck up

The4codeblocks committed:
stop with the manually constructed vector

The4codeblocks committed:
shut luacheck up

The4codeblocks commented:
so....?

The4codeblocks committed:
fix crash

The4codeblocks committed:
forgot a paren

The4codeblocks commented:
any news here?

The4codeblocks committed:
Merge branch 'master' into patch-1

The4codeblocks commented:
?

The4codeblocks commented:
?

The4codeblocks committed:
Merge branch 'master' into patch-1

wsor4035 commented on crafts.lua:
```diff
@@ -316,3 +316,51 @@ if pipeworks.enable_node_breaker then
+	buckets.air = "bucket:bucket_empty"
+	buckets.water = "bucket:bucket_water"
+	buckets.river_water = "bucket:bucket_river_water"
+	buckets.lava = "bucket:bucket_lava"
+elseif core.get_modpath("mcl_buckets") then
+	buckets.air = "mcl_buckets:bucket_empty"
+	buckets.water = "mcl_buckets:bucket_water"
+	buckets.river_water = "mcl_buckets:bucket_river_water"
+	buckets.lava = "mcl_buckets:bucket_lava"
+end
```
this should come from xcompat

The4codeblocks replied:
should be done
(riverwater & lava buckets don't have xcompat entries)

wsor4035 commented:
i have only skimmed this code, but would it be possible to break parts of this pr off and merge them in? as this is quite huge

SwissalpS commented on common.lua:
```diff
@@ -211,19 +211,22 @@ function fs_helpers.cycling_button(meta, base, meta_name, values)
 		local size, spacing = 1, 0.1
 		local hotbar_len = i3.settings.hotbar_len or (i3.settings.legacy_inventory and 8 or 9)
-		local inv_size = i3.settings.inv_size or (hotbar_len * 4)
+		local hotbar_width = (i3.settings.inv_size and (i3.settings.inv_size/hotbar_len)) or 4
```
spaces around operators, please

SwissalpS replied:
Not only here, in general wherever lines have been changed also improve layout. Thanks.

The4codeblocks committed:
readability

The4codeblocks committed:
xcompat where possible

oddly enough, xcompat doesn't do anything with riverwater and lava

The4codeblocks committed:
announce deprecation

"Deprecate flowing_logic.lua for classic flow logic
Mark the file as deprecated due to new flow logic requirements."
- copilot

The4codeblocks committed:
reduce loading baggage

SwissalpS commented on init.lua:
```diff
@@ -81,7 +81,7 @@ dofile(pipeworks.modpath.."/autoplace_pipes.lua")
 dofile(pipeworks.modpath.."/luaentity.lua")
 dofile(pipeworks.modpath.."/item_transport.lua")
-dofile(pipeworks.modpath.."/flowing_logic.lua")
+if pipeworks.toggles.pipe_mode == "classic" then dofile(pipeworks.modpath.."/flowing_logic.lua") end
```
Please take a look at the code style in similar cases e.g. lines 97 to 105.

The4codeblocks committed:
fix styling

The4codeblocks commented:
so what

The4codeblocks committed:
update

The4codeblocks committed:
fix luacheck

The4codeblocks committed:
translation

The4codeblocks committed:
more translation, tooltip

The4codeblocks committed:
namespace fluid information fields, unify field name definition a little

The4codeblocks committed:
Merge branch 'master' into patch-1

The4codeblocks committed:
avoid water dependency when water is absent, use dummy texture instead

The4codeblocks committed:
Merge remote-tracking branch 'origin/patch-1' into patch-1

The4codeblocks committed:
Merge branch 'master' into patch-1

The4codeblocks committed:
Merge remote-tracking branch 'parent/master' into patch-1

# Conflicts:
#	autocrafter.lua
#	trashcan.lua

The4codeblocks committed:
reimplement fluid crafting for autocrafter redo

The4codeblocks committed:
unused local

The4codeblocks committed:
missed the translator

The4codeblocks committed:
missed the unneeded init

The4codeblocks committed:
Merge remote-tracking branch 'parent/master' into patch-1

# Conflicts:
#	locale/template.txt
#	mod.conf

The4codeblocks committed:
add gitignore, follow translation migration
